thread/synchronization/thread_synchronization.odin

```diff
@@ -0,0 +1,145 @@
+//
+//  Demonstrates how to spawn multiple threads and safely access data from each.
+//
+package thread_sync_example
+
+import "core:fmt"
+import "core:math/rand"
+import "core:sync"
+import "core:thread"
+import "core:time"
+
+Work_Item :: struct {
+	item_tag:        i32,
+	processing_time: f32,
+}
+
+create_randomized_queue :: proc(num_items: int) -> (q: [dynamic]Work_Item) {
+	// This initializes the queue with a length of zero, and a capacity of `num_items`.
+	// Pre-allocating space when you know what you need is good!
+	q = make([dynamic]Work_Item, 0, num_items)
+
+	// Here we loop from 0 to the number of items and add an item to our queue.
+	for i in 0 ..< num_items {
+		item: Work_Item
+		item.item_tag = i32(i) + 1
+		// This sets the item's processing time to a value between 0.1 and 1.0 (exclusive as well).
+		item.processing_time = rand.float32_range(0.1, 0.51)
+		append(&q, item)
+	}
+
+	return
+}
+
+// This is the procedure that we'll be running in the threads that we spawn later.
+process_item :: proc(queue: ^[dynamic]Work_Item, mutex: ^sync.Mutex, thread_identifier: int) {
+	// This proc is essentially an infinite loop that breaks once it no longer has any data to process.
+
+	// We should always call `mutex_unlock` at the end of the thread proc so that we don't accidentally keep it locked.
+	// Since unlocking an unlocked mutex is a no-op anyway, it's fine to do this even if we don't need to.
+	defer sync.mutex_unlock(mutex)
+
+	// This loop is infinite.
+	loop: for {
+		// First we need to try to get a lock on our mutex. 
+		// That way we know whether we can safely access our queue, or if another thread is using it already.
+		if sync.mutex_try_lock(mutex) {
```

Comment:
Okay, I just noticed this one... I don't think it's a good idea to use `try_lock` in a synchronization example, because for this scenario it is a bad idea and you lose out on performance by waiting for more than you need to. It's better to show how critical sections work, which is a more general, and a better way to do synchronization, than burning CPU cycles on a loop.